File: web/src/sim/military/Stance.ts
import { divF32, f32, mulF32, subF32 } from '~/sim/determinism/float32'
import { roundHalfToEven } from '~/sim/determinism/rounding'
import type { Army } from '~/sim/military/Army'
import { SupplyStatus, type SupplySystem } from '~/sim/military/SupplySystem'
import { HOURS_PER_DAY } from '~/sim/time/GameDate'
import { favoursDefence, Terrain } from '~/sim/world/Terrain'
import type { WorldState } from '~/sim/world/WorldState'

export enum StanceKind {
  Assault = 0,
  Hold = 1,
  Ambush = 2,
  Siege = 3,
  Screen = 4,
  Raid = 5,
}

export const REORGANISATION_TICKS = 6

export const ASSAULT_ATTACK = f32(1.15)
export const ASSAULT_DAMAGE_TAKEN = f32(1.1)
export const ASSAULT_ROUGH_TERRAIN_ATTACK = f32(0.85)

export const SCREEN_ATTACK = f32(0.7)
export const SCREEN_DAMAGE_TAKEN = f32(0.8)
export const SCREEN_WITHDRAWAL_HEALTH = f32(0.6)

export const RAID_ATTACK = f32(0.8)
export const RAID_SPEED = f32(1.2)
export const RAID_SUPPLY_PENALTY = f32(0.5)
export const RAID_SUPPLY_TICKS = 72

export const DAILY_ENTRENCHMENT = f32(0.1)
export const AMBUSH_ENTRENCHMENT_CAP = f32(0.5)
export const ENTRENCHED_CITY_DAMAGE_TAKEN = f32(0.75)
export const ENTRENCHED_FIELD_DAMAGE_TAKEN = f32(0.9)

export const CITY_ASSAULT_PENALTY = f32(0.75)
export const ASSAULT_CITY_PENALTY = f32(0.85)

/* Pengintaian yang menyelesaikan sebuah provinsi menjadi perkiraan atau lebih
   baik menembus penyergapan; apa pun yang lebih kabur dari itu tidak. */
export const HIDDEN_INTEL_THRESHOLD = 2

/* Padanan int.MaxValue: batas atas C# saat memotong selisih tick yang bertipe
   long menjadi jumlah jam yang bertipe int. */
const INT32_MAX = 2147483647

export function attackMultiplier(kind: StanceKind): number {
  switch (kind) {
    case StanceKind.Assault:
      return ASSAULT_ATTACK
    case StanceKind.Screen:
      return SCREEN_ATTACK
    case StanceKind.Raid:
      return RAID_ATTACK
    default:
      return 1
  }
}

export function damageTakenMultiplier(kind: StanceKind): number {
  switch (kind) {
    case StanceKind.Assault:
      return ASSAULT_DAMAGE_TAKEN
    case StanceKind.Screen:
      return SCREEN_DAMAGE_TAKEN
    default:
      return 1
  }
}

export function speedMultiplier(kind: StanceKind): number {
  return kind === StanceKind.Raid ? RAID_SPEED : 1
}

/* Menyerbu celah gunung atau blok padat dengan sikap assault menghabiskan
   bonusnya dan lebih: berat tambahannya tidak punya tempat untuk digelar. */
export function terrainAttackModifier(kind: StanceKind, terrain: Terrain): number {
  return kind === StanceKind.Assault &&
    (terrain === Terrain.Mountains || terrain === Terrain.Urban)
    ? ASSAULT_ROUGH_TERRAIN_ATTACK
    : 1
}

export function cityAssaultMultiplier(kind: StanceKind): number {
  return kind === StanceKind.Assault ? ASSAULT_CITY_PENALTY : CITY_ASSAULT_PENALTY
}

export function entrenchmentCap(kind: StanceKind): number {
  switch (kind) {
    case StanceKind.Hold:
      return 1
    case StanceKind.Ambush:
      return AMBUSH_ENTRENCHMENT_CAP
    default:
      return 0
  }
}

export function maxDugHours(kind: StanceKind): number {
  return Math.trunc(
    roundHalfToEven(mulF32(divF32(entrenchmentCap(kind), DAILY_ENTRENCHMENT), HOURS_PER_DAY)),
  )
}

export function entrenchmentFrom(kind: StanceKind, dugHours: number): number {
  return Math.min(
    entrenchmentCap(kind),
    divF32(mulF32(dugHours, DAILY_ENTRENCHMENT), HOURS_PER_DAY),
  )
}

export function entrenchmentDamageTaken(entrenchment: number, inCity: boolean): number {
  if (entrenchment <= 0) return 1

  const lowest = inCity ? ENTRENCHED_CITY_DAMAGE_TAKEN : ENTRENCHED_FIELD_DAMAGE_TAKEN
  return subF32(1, mulF32(entrenchment, subF32(1, lowest)))
}

export function entrenches(kind: StanceKind): boolean {
  return kind === StanceKind.Hold || kind === StanceKind.Ambush
}

export function canCapture(kind: StanceKind): boolean {
  return kind !== StanceKind.Raid
}

/* Tidak bisa bergerak adalah izin, bukan kecepatan nol: tumpukan yang menggali
   dan lalu kocar-kacir atau diperintahkan keluar tetap harus menempuh jalan
   dengan laju yang bisa dipakai. */
export function canMove(kind: StanceKind): boolean {
  return kind !== StanceKind.Hold && kind !== StanceKind.Ambush
}

export function engagesEnemyArmies(kind: StanceKind): boolean {
  return kind !== StanceKind.Siege
}

export function advancesSiege(kind: StanceKind): boolean {
  return kind === StanceKind.Siege
}

/* Menggali lebih cepat di tanah yang sudah memberi awalan: garis pepohonan atau
   deretan rumah sudah setengah kedudukan sebelum ada yang mengangkat sekop. */
export function diggingRateIn(terrain: Terrain): number {
  return terrain === Terrain.Forest || terrain === Terrain.Hills || terrain === Terrain.Urban
    ? 2
    : 1
}

/*
 * Parit disimpan sebagai jam yang benar-benar dipakai menggali, bukan sebagai
 * pecahan yang tumbuh tiap tick. Menurunkan tingkatnya dari bilangan bulat
 * menjaga laju menggali yang berlipat dan sikap yang dibatasi agar tidak
 * menyimpang sepanjang ratusan tick yang dilewati garis yang menggali.
 */
export class StanceState {
  dugHours = 0

  constructor(
    public army: Army,
    public kind: StanceKind,
    public terrain: Terrain,
    public changedAtTick: number,
  ) {}

  get armyId(): number {
    return this.army.id
  }

  get entrenchment(): number {
    return entrenchmentFrom(this.kind, this.dugHours)
  }
}

export interface StanceOrder {
  readonly accepted: boolean
  readonly reason: string
}

const ACCEPTED: StanceOrder = { accepted: true, reason: '' }

/*
 * Enam sikap yang bisa diambil sebuah tumpukan, dan ongkos berubah pikiran.
 * Sikap baru berlaku setelah jendela reorganisasi, sehingga ia adalah komitmen
 * yang dibuat sebelum pertempuran alih-alih tuas yang ditarik setelah tembakan
 * dimulai; selama enam tick itu tumpukan bertempur tanpa satu pun keunggulan
 * sikapnya dan tanpa satu pun kerugiannya.
 *
 * Parit sengaja berada di luar aturan itu. Parit adalah tanah yang sudah
 * digali, bukan niat, sehingga ia tetap melindungi tumpukan selama ia
 * berorganisasi ulang dan baru hilang ketika tumpukan meninggalkannya.
 */
export class StanceSystem {
  readonly #states = new Map<number, StanceState>()
  readonly #ordered: StanceState[] = []
  readonly #raidedUntil = new Map<number, number>()

  #lastTick: number

  constructor(
    private readonly world: WorldState,
    private readonly supply: SupplySystem,
  ) {
    this.#lastTick = world.clock.tick
  }

  get trackedArmies(): number {
    return this.#ordered.length
  }

  stateOf(armyId: number): StanceState | null {
    return this.#states.get(armyId) ?? null
  }

  stanceOf(armyId: number): StanceKind | null {
    return this.stateOf(armyId)?.kind ?? null
  }

  effectiveStanceOf(armyId: number): StanceKind | null {
    const state = this.stateOf(armyId)
    return state === null || this.#isReorganising(state) ? null : state.kind
  }

  isReorganising(armyId: number): boolean {
    const state = this.stateOf(armyId)
    return state !== null && this.#isReorganising(state)
  }

  attackMultiplierFor(armyId: number, terrain?: Terrain): number {
    const kind = this.effectiveStanceOf(armyId)
    if (kind === null) return 1

    return terrain === undefined
      ? attackMultiplier(kind)
      : mulF32(attackMultiplier(kind), terrainAttackModifier(kind, terrain))
  }

  damageTakenMultiplierFor(armyId: number): number {
    const state = this.stateOf(armyId)
    if (state === null) return 1

    const stance = this.#isReorganising(state) ? 1 : damageTakenMultiplier(state.kind)
    return mulF32(stance, this.#entrenchmentDamageTakenFor(state))
  }

  speedMultiplierFor(armyId: number): number {
    const kind = this.effectiveStanceOf(armyId)
    return kind === null ? 1 : speedMultiplier(kind)
  }

  cityAssaultMultiplierFor(armyId: number): number {
    const kind = this.effectiveStanceOf(armyId)
    return kind === null ? CITY_ASSAULT_PENALTY : cityAssaultMultiplier(kind)
  }

  entrenchmentOf(armyId: number): number {
    return this.stateOf(armyId)?.entrenchment ?? 0
  }

  canCapture(armyId: number): boolean {
    const kind = this.effectiveStanceOf(armyId)
    return kind === null || canCapture(kind)
  }

  canMove(armyId: number): boolean {
    const kind = this.effectiveStanceOf(armyId)
    return kind === null || canMove(kind)
  }

  engagesEnemyArmies(armyId: number): boolean {
    const kind = this.effectiveStanceOf(armyId)
    return kind === null || engagesEnemyArmies(kind)
  }

  advancesSiege(armyId: number): boolean {
    return this.effectiveStanceOf(armyId) === StanceKind.Siege
  }

  retreatsWithoutPenalty(armyId: number): boolean {
    return this.effectiveStanceOf(armyId) === StanceKind.Screen
  }

  shouldWithdraw(armyId: number): boolean {
    const state = this.stateOf(armyId)

    return (
      state !== null &&
      !this.#isReorganising(state) &&
      state.kind === StanceKind.Screen &&
      state.army.health < SCREEN_WITHDRAWAL_HEALTH
    )
  }

  isHiddenFrom(armyId: number, enemyIntelLevel: number): boolean {
    const state = this.stateOf(armyId)

    return (
      state !== null &&
      !this.#isReorganising(state) &&
      state.kind === StanceKind.Ambush &&
      enemyIntelLevel < HIDDEN_INTEL_THRESHOLD
    )
  }

  isRaided(province: number): boolean {
    const until = this.#raidedUntil.get(province)
    return until !== undefined && until > this.world.clock.tick
  }

  raidedSupplyMultiplierOf(province: number): number {
    return this.isRaided(province) ? RAID_SUPPLY_PENALTY : 1
  }

  set(army: Army, kind: StanceKind, terrain: Terrain): StanceOrder {
    const refusal = this.#refusalFor(army, kind, terrain)
    if (refusal !== null) return { accepted: false, reason: refusal }

    const now = this.world.clock.tick
    const state = this.#states.get(army.id)

    if (state !== undefined) {
      state.army = army
      state.terrain = terrain

      /* Mengulang perintah yang sedang berlaku bukan perubahan pikiran,
         sehingga ia tidak boleh mengulang reorganisasi maupun penggaliannya. */
      if (state.kind === kind) return ACCEPTED

      state.kind = kind
      state.changedAtTick = now
      state.dugHours = Math.min(state.dugHours, maxDugHours(kind))
      return ACCEPTED
    }

    const created = new StanceState(army, kind, terrain, now)
    this.#states.set(army.id, created)
    this.#ordered.push(created)
    return ACCEPTED
  }

  onArmyMoved(armyId: number): void {
    const state = this.#states.get(armyId)
    if (state !== undefined) state.dugHours = 0
  }

  forget(armyId: number): void {
    const state = this.#states.get(armyId)
    if (state === undefined) return

    this.#states.delete(armyId)

    const index = this.#ordered.indexOf(state)
    if (index >= 0) this.#ordered.splice(index, 1)
  }

  /* Tidak ada yang terjadi dua kali dalam satu jam permainan: jam, bukan jumlah
     panggilan, yang menentukan seberapa banyak sebuah tumpukan sudah menggali. */
  tick(): void {
    const now = this.world.clock.tick

    if (now <= this.#lastTick) {
      this.#lastTick = now
      return
    }

    const hours = Math.min(now - this.#lastTick, INT32_MAX)
    this.#lastTick = now

    this.#dropDestroyed()
    this.#expireRaids(now)

    for (const state of this.#ordered) {
      if (this.#isReorganising(state)) continue

      this.#dig(state, hours)
      this.#raid(state, now)
    }
  }

  #isReorganising(state: StanceState): boolean {
    return this.world.clock.tick - state.changedAtTick < REORGANISATION_TICKS
  }

  #entrenchmentDamageTakenFor(state: StanceState): number {
    return entrenchmentDamageTaken(
      state.entrenchment,
      this.world.provinces.isCity[state.army.province] !== 0,
    )
  }

  #dig(state: StanceState, hours: number): void {
    if (!entrenches(state.kind) || !this.supply.canEntrench(state.army.province)) return

    const cap = maxDugHours(state.kind)
    if (state.dugHours >= cap) return

    state.dugHours = Math.min(cap, state.dugHours + hours * diggingRateIn(state.terrain))
  }

  #raid(state: StanceState, now: number): void {
    if (state.kind !== StanceKind.Raid) return

    const province = state.army.province
    if (this.world.provinces.controller[province] === state.army.nation) return

    this.#raidedUntil.set(province, now + RAID_SUPPLY_TICKS)
  }

  /* Kunci yang kedaluwarsa dikumpulkan dulu lalu dihapus: himpunannya tidak
     bergantung pada urutan kunjungan, sehingga peta tetap boleh diiterasi
     menurut urutan sisipnya. */
  #expireRaids(now: number): void {
    if (this.#raidedUntil.size === 0) return

    const expired: number[] = []
    for (const [province, until] of this.#raidedUntil) {
      if (until <= now) expired.push(province)
    }

    for (const province of expired) this.#raidedUntil.delete(province)
  }

  #dropDestroyed(): void {
    for (let i = this.#ordered.length - 1; i >= 0; i--) {
      const state = this.#ordered[i]!
      if (!state.army.isDestroyed) continue

      this.#ordered.splice(i, 1)
      this.#states.delete(state.armyId)
    }
  }

  #refusalFor(army: Army, kind: StanceKind, terrain: Terrain): string | null {
    if (army.isDestroyed) return `Army ${army.id} has been destroyed.`

    if (
      kind === StanceKind.Assault &&
      this.supply.effectiveStatusOf(army.province) === SupplyStatus.CutOff
    ) {
      return `Army ${army.id} is cut off and cannot mount an assault.`
    }

    /* Perlindungan yang dibutuhkan penyergapan adalah tanah yang sama yang
       sudah memihak pembela, sehingga aturannya dibaca dari satu tempat
       alih-alih dieja dua kali lalu menyimpang. */
    if (kind === StanceKind.Ambush && !favoursDefence(terrain)) {
      return `Ambush needs rough terrain; ${Terrain[terrain]} leaves the stack in the open.`
    }

    if (kind === StanceKind.Ambush && !this.#isSupplied(army.province)) {
      return `Ambush needs a supplied position; army ${army.id} has none.`
    }

    if (kind === StanceKind.Siege && !this.#isSupplied(army.province)) {
      return `A siege cannot be maintained on broken supply by army ${army.id}.`
    }

    return null
  }

  #isSupplied(province: number): boolean {
    return this.supply.effectiveStatusOf(province) === SupplyStatus.Supplied
  }
}

File: sim/NationRise.Core/Military/Stance.cs
using NationRise.Core.Time;
using NationRise.Core.World;

namespace NationRise.Core.Military;

public enum StanceKind : byte
{
    Assault = 0,
    Hold = 1,
    Ambush = 2,
    Siege = 3,
    Screen = 4,
    Raid = 5,
}

/*
   Entrenchment is stored as hours actually spent digging rather than as a
   fraction that grows every tick. Deriving the level from a whole number keeps
   a doubled digging rate and a capped stance from drifting apart over the
   hundreds of ticks a dug-in line survives.
*/
public sealed class StanceState
{
    internal StanceState(Army army, StanceKind kind, Terrain terrain, long tick)
    {
        Army = army;
        Kind = kind;
        Terrain = terrain;
        ChangedAtTick = tick;
    }

    public Army Army { get; internal set; }
    public StanceKind Kind { get; internal set; }
    public Terrain Terrain { get; internal set; }
    public long ChangedAtTick { get; internal set; }
    public int DugHours { get; internal set; }

    public int ArmyId => Army.Id;

    public float Entrenchment => StanceSystem.EntrenchmentOf(Kind, DugHours);
}

/*
   Six postures a stack can adopt, and the cost of changing its mind. A stance
   only takes hold after a reorganisation window, so posture is a commitment
   made ahead of a fight rather than a lever pulled once the shooting starts;
   for those six ticks the stack fights with none of the stance's advantages
   and none of its drawbacks.

   Entrenchment is deliberately outside that rule. Trenches are ground that has
   been dug, not an intention, so they keep protecting the stack while it
   reorganises and vanish only when it leaves them behind.
*/
public sealed class StanceSystem
{
    public const int ReorganisationTicks = 6;

    public const float AssaultAttack = 1.15f;
    public const float AssaultDamageTaken = 1.10f;
    public const float AssaultRoughTerrainAttack = 0.85f;

    public const float ScreenAttack = 0.70f;
    public const float ScreenDamageTaken = 0.80f;
    public const float ScreenWithdrawalHealth = 0.60f;

    public const float RaidAttack = 0.80f;
    public const float RaidSpeed = 1.20f;
    public const float RaidSupplyPenalty = 0.50f;
    public const int RaidSupplyTicks = 72;

    public const float DailyEntrenchment = 0.10f;
    public const float AmbushEntrenchmentCap = 0.50f;
    public const float EntrenchedCityDamageTaken = 0.75f;
    public const float EntrenchedFieldDamageTaken = 0.90f;

    public const float CityAssaultPenalty = 0.75f;
    public const float AssaultCityPenalty = 0.85f;

    /* Recon that resolves a province to an estimate or better sees through an
       ambush; anything vaguer than that does not. */
    public const int HiddenIntelThreshold = 2;

    private readonly WorldState _world;
    private readonly SupplySystem _supply;

    private readonly Dictionary<int, StanceState> _states = [];
    private readonly List<StanceState> _ordered = [];
    private readonly Dictionary<int, long> _raidedUntil = [];

    private long _lastTick;

    public StanceSystem(WorldState world, SupplySystem supply)
    {
        ArgumentNullException.ThrowIfNull(world);
        ArgumentNullException.ThrowIfNull(supply);

        _world = world;
        _supply = supply;
        _lastTick = world.Clock.Tick;
    }

    public int TrackedArmies => _ordered.Count;

    public static float AttackMultiplier(StanceKind kind) => kind switch
    {
        StanceKind.Assault => AssaultAttack,
        StanceKind.Screen => ScreenAttack,
        StanceKind.Raid => RaidAttack,
        _ => 1f,
    };

    public static float DamageTakenMultiplier(StanceKind kind) => kind switch
    {
        StanceKind.Assault => AssaultDamageTaken,
        StanceKind.Screen => ScreenDamageTaken,
        _ => 1f,
    };

    public static float SpeedMultiplier(StanceKind kind) =>
        kind == StanceKind.Raid ? RaidSpeed : 1f;

    /* Storming a mountain pass or a built-up block with an assault stance
       spends the bonus and more: the extra weight has nowhere to deploy. */
    public static float TerrainAttackModifier(StanceKind kind, Terrain terrain) =>
        kind == StanceKind.Assault && terrain is Terrain.Mountains or Terrain.Urban
            ? AssaultRoughTerrainAttack
            : 1f;

    public static float CityAssaultMultiplier(StanceKind kind) =>
        kind == StanceKind.Assault ? AssaultCityPenalty : CityAssaultPenalty;

    public static float EntrenchmentCap(StanceKind kind) => kind switch
    {
        StanceKind.Hold => 1f,
        StanceKind.Ambush => AmbushEntrenchmentCap,
        _ => 0f,
    };

    public static int MaxDugHours(StanceKind kind) =>
        (int)MathF.Round(EntrenchmentCap(kind) / DailyEntrenchment * GameDate.HoursPerDay);

    public static float EntrenchmentOf(StanceKind kind, int dugHours) => MathF.Min(
        EntrenchmentCap(kind),
        dugHours * DailyEntrenchment / GameDate.HoursPerDay);

    public static float EntrenchmentDamageTaken(float entrenchment, bool inCity)
    {
        if (entrenchment <= 0f)
        {
            return 1f;
        }

        float floor = inCity ? EntrenchedCityDamageTaken : EntrenchedFieldDamageTaken;
        return 1f - entrenchment * (1f - floor);
    }

    public static bool Entrenches(StanceKind kind) =>
        kind is StanceKind.Hold or StanceKind.Ambush;

    public static bool CanCapture(StanceKind kind) => kind != StanceKind.Raid;

    /* Immobility is a permission, not a speed of zero: a dug-in stack that is
       routed or ordered out still has to travel at a usable rate. */
    public static bool CanMove(StanceKind kind) =>
        kind is not (StanceKind.Hold or StanceKind.Ambush);

    public static bool EngagesEnemyArmies(StanceKind kind) => kind != StanceKind.Siege;

    public static bool AdvancesSiege(StanceKind kind) => kind == StanceKind.Siege;

    /* Digging is quicker where the ground already offers a start: a wood line
       or a row of houses is half a position before anyone lifts a shovel. */
    public static int DiggingRateIn(Terrain terrain) =>
        terrain is Terrain.Forest or Terrain.Hills or Terrain.Urban ? 2 : 1;

    public StanceState? StateOf(int armyId) =>
        _states.TryGetValue(armyId, out StanceState? state) ? state : null;

    public StanceKind? StanceOf(int armyId) => StateOf(armyId)?.Kind;

    public StanceKind? EffectiveStanceOf(int armyId)
    {
        StanceState? state = StateOf(armyId);
        return state is null || IsReorganising(state) ? null : state.Kind;
    }

    public bool IsReorganising(int armyId)
    {
        StanceState? state = StateOf(armyId);
        return state is not null && IsReorganising(state);
    }

    public float AttackMultiplierFor(int armyId)
    {
        StanceKind? kind = EffectiveStanceOf(armyId);
        return kind is null ? 1f : AttackMultiplier(kind.Value);
    }

    public float AttackMultiplierFor(int armyId, Terrain terrain)
    {
        StanceKind? kind = EffectiveStanceOf(armyId);
        return kind is null
            ? 1f
            : AttackMultiplier(kind.Value) * TerrainAttackModifier(kind.Value, terrain);
    }

    public float DamageTakenMultiplierFor(int armyId)
    {
        StanceState? state = StateOf(armyId);
        if (state is null)
        {
            return 1f;
        }

        float stance = IsReorganising(state) ? 1f : DamageTakenMultiplier(state.Kind);
        return stance * EntrenchmentDamageTakenFor(state);
    }

    public float SpeedMultiplierFor(int armyId)
    {
        StanceKind? kind = EffectiveStanceOf(armyId);
        return kind is null ? 1f : SpeedMultiplier(kind.Value);
    }

    public float CityAssaultMultiplierFor(int armyId)
    {
        StanceKind? kind = EffectiveStanceOf(armyId);
        return kind is null ? CityAssaultPenalty : CityAssaultMultiplier(kind.Value);
    }

    public float EntrenchmentOf(int armyId) => StateOf(armyId)?.Entrenchment ?? 0f;

    public bool CanCapture(int armyId)
    {
        StanceKind? kind = EffectiveStanceOf(armyId);
        return kind is null || CanCapture(kind.Value);
    }

    public bool CanMove(int armyId)
    {
        StanceKind? kind = EffectiveStanceOf(armyId);
        return kind is null || CanMove(kind.Value);
    }

    public bool EngagesEnemyArmies(int armyId)
    {
        StanceKind? kind = EffectiveStanceOf(armyId);
        return kind is null || EngagesEnemyArmies(kind.Value);
    }

    public bool AdvancesSiege(int armyId) => EffectiveStanceOf(armyId) == StanceKind.Siege;

    public bool RetreatsWithoutPenalty(int armyId) => EffectiveStanceOf(armyId) == StanceKind.Screen;

    public bool ShouldWithdraw(int armyId)
    {
        StanceState? state = StateOf(armyId);
        return state is not null
            && !IsReorganising(state)
            && state.Kind == StanceKind.Screen
            && state.Army.Health < ScreenWithdrawalHealth;
    }

    public bool IsHiddenFrom(int armyId, int enemyIntelLevel)
    {
        StanceState? state = StateOf(armyId);
        return state is not null
            && !IsReorganising(state)
            && state.Kind == StanceKind.Ambush
            && enemyIntelLevel < HiddenIntelThreshold;
    }

    public bool IsRaided(int province) =>
        _raidedUntil.TryGetValue(province, out long until) && until > _world.Clock.Tick;

    public float RaidedSupplyMultiplierOf(int province) =>
        IsRaided(province) ? RaidSupplyPenalty : 1f;

    public bool Set(Army army, StanceKind kind, Terrain terrain, out string reason)
    {
        ArgumentNullException.ThrowIfNull(army);

        if (!IsPermitted(army, kind, terrain, out reason))
        {
            return false;
        }

        long now = _world.Clock.Tick;

        if (_states.TryGetValue(army.Id, out StanceState? state))
        {
            state.Army = army;
            state.Terrain = terrain;

            /* Re-issuing the standing order is not a change of mind, so it
               must not restart the reorganisation or the digging. */
            if (state.Kind == kind)
            {
                return true;
            }

            state.Kind = kind;
            state.ChangedAtTick = now;
            state.DugHours = Math.Min(state.DugHours, MaxDugHours(kind));
            return true;
        }

        var created = new StanceState(army, kind, terrain, now);
        _states[army.Id] = created;
        _ordered.Add(created);
        return true;
    }

    public void OnArmyMoved(int armyId)
    {
        if (_states.TryGetValue(armyId, out StanceState? state))
        {
            state.DugHours = 0;
        }
    }

    public void Forget(int armyId)
    {
        if (_states.Remove(armyId, out StanceState? state))
        {
            _ordered.Remove(state);
        }
    }

    /* Nothing happens twice in the same game hour: the clock, not the number of
       calls, decides how much digging a stack has done. */
    public void Tick()
    {
        long now = _world.Clock.Tick;

        if (now <= _lastTick)
        {
            _lastTick = now;
            return;
        }

        int hours = (int)Math.Min(now - _lastTick, int.MaxValue);
        _lastTick = now;

        DropDestroyed();
        ExpireRaids(now);

        foreach (StanceState state in _ordered)
        {
            if (IsReorganising(state))
            {
                continue;
            }

            Dig(state, hours);
            Raid(state, now);
        }
    }

    private bool IsReorganising(StanceState state) =>
        _world.Clock.Tick - state.ChangedAtTick < ReorganisationTicks;

    private float EntrenchmentDamageTakenFor(StanceState state) => EntrenchmentDamageTaken(
        state.Entrenchment,
        _world.Provinces.IsCity[state.Army.Province]);

    private void Dig(StanceState state, int hours)
    {
        if (!Entrenches(state.Kind) || !_supply.CanEntrench(state.Army.Province))
        {
            return;
        }

        int cap = MaxDugHours(state.Kind);
        if (state.DugHours >= cap)
        {
            return;
        }

        state.DugHours = Math.Min(cap, state.DugHours + hours * DiggingRateIn(state.Terrain));
    }

    private void Raid(StanceState state, long now)
    {
        if (state.Kind != StanceKind.Raid)
        {
            return;
        }

        int province = state.Army.Province;
        if (_world.Provinces.Controller[province] == state.Army.Nation)
        {
            return;
        }

        _raidedUntil[province] = now + RaidSupplyTicks;
    }

    private void ExpireRaids(long now)
    {
        if (_raidedUntil.Count == 0)
        {
            return;
        }

        List<int> expired = [];
        foreach ((int province, long until) in _raidedUntil)
        {
            if (until <= now)
            {
                expired.Add(province);
            }
        }

        foreach (int province in expired)
        {
            _raidedUntil.Remove(province);
        }
    }

    private void DropDestroyed()
    {
        for (int i = _ordered.Count - 1; i >= 0; i--)
        {
            StanceState state = _ordered[i];
            if (!state.Army.IsDestroyed)
            {
                continue;
            }

            _ordered.RemoveAt(i);
            _states.Remove(state.ArmyId);
        }
    }

    private bool IsPermitted(Army army, StanceKind kind, Terrain terrain, out string reason)
    {
        reason = string.Empty;

        if (army.IsDestroyed)
        {
            reason = $"Army {army.Id} has been destroyed.";
            return false;
        }

        if (kind == StanceKind.Assault
            && _supply.EffectiveStatusOf(army.Province) == SupplyStatus.CutOff)
        {
            reason = $"Army {army.Id} is cut off and cannot mount an assault.";
            return false;
        }

        /* The cover an ambush needs is the same ground that already favours a
           defender, so the rule is read from one place instead of being spelled
           out twice and drifting. */
        if (kind == StanceKind.Ambush && !terrain.FavoursDefence())
        {
            reason = $"Ambush needs rough terrain; {terrain} leaves the stack in the open.";
            return false;
        }

        if (kind == StanceKind.Ambush && !IsSupplied(army.Province))
        {
            reason = $"Ambush needs a supplied position; army {army.Id} has none.";
            return false;
        }

        if (kind == StanceKind.Siege && !IsSupplied(army.Province))
        {
            reason = $"A siege cannot be maintained on broken supply by army {army.Id}.";
            return false;
        }

        return true;
    }

    private bool IsSupplied(int province) =>
        _supply.EffectiveStatusOf(province) == SupplyStatus.Supplied;
}
